Clean up the kubeadm high availability instructions. Fixes Issue 15284

rcythr opened the pull request:
Fixes #15284 

/assign @fabriziopandini @neolit123

rcythr commented:
/sig cluster-lifecycle

kbhawkey commented:
hello @rcythr , would you coordinate with PR #15372. Thanks!

fabriziopandini commented on content/en/docs/setup/production-environment/tools/kubeadm/high-availability.md:
```diff
@@ -58,32 +58,13 @@ For the external etcd cluster only, you also need:
 
-    - Make sure the address of the load balancer always matches
-      the address of kubeadm's `ControlPlaneEndpoint`.
+  - Check if your deployment environment has a highly available load balancer, and use that if possible. Otherwise, setup a TCP forwarding load balancer (such as [HAProxy](http://www.haproxy.org/)). Make sure to follow the documentation to properly setup this load balancer in a highly-available configuration. Otherwise, your load balancer becomes a single point of failure.
```
After re-reading, I think this sentence should become the first point (first check what you have, second create). Also, plese change _"Otherwise, setup a TCP forwarding load balancer (such as [HAProxy](http://www.haproxy.org/))"_ into Otherwise, _"select a TCP forwarding load balancer such as [HAProxy](http://www.haproxy.org/)"_, in order to be more in line with the "check what you have" message

tengqm commented:
@rcythr Feedback reasonable to you?

rcythr commented:
@fabriziopandini How about this? Sorry for the delay, I've been swamped this week. 
 
@tengqm Thanks for the reminder.

fabriziopandini commented:
@rcythr thanks for the updates! I'm ok with the changes, I leave to @neolit123 giving the final green light

rcythr commented:
/assign @kbarnard10

rcythr commented:
/hold

In light of https://github.com/kubernetes/kubeadm/issues/1685 we should revise our advice to be more explicit about what kind of things the LB needs to be able to do. Azure's LB currently cannot fulfil those requirements.

rcythr commented:
@neolit123 @fabriziopandini I modified the text to make the HA LB a requirement instead of trying to provide steps for setting it up -- we can't get into details about the specific implementations. I added a note about hairpin NAT in light of kubernetes/kubeadm#1685

What do you think?

rcythr commented:
/hold cancel
I have adjusted this to incorporate content from kubernetes/kubeadm#1685.

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/high-availability.md:
```diff
@@ -40,6 +40,7 @@ For both methods you need this infrastructure:
   the masters
 - Three machines that meet [kubeadm's minimum
   requirements](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#before-you-begin) for the workers
+- A highly available load balancer that operates at Layer 4 (such as one provided by a cloud provider, or [HAProxy](http://www.haproxy.org/)) with it's own name in DNS. The load balancer should be configured to forward the apiserver port (default: 6443) on the load balancer to the apiserver port on all control plane nodes. Before using a cloud provider provided load balancer, double check the doc ...
```
we should probably mention HTTPS and health checks here.
@ereslibre @fabriziopandini how would you word that?

ereslibre replied:
I would say that even if it's a layer 4 LB it should have the ability to perform HTTPS health checks (even if the certificate chain is ignored). That way it will remove from the active pool unhealthy apiservers.

ereslibre commented on content/en/docs/setup/production-environment/tools/kubeadm/high-availability.md:
```diff
@@ -40,6 +40,7 @@ For both methods you need this infrastructure:
   the masters
 - Three machines that meet [kubeadm's minimum
   requirements](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#before-you-begin) for the workers
+- A highly available load balancer that operates at Layer 4 (such as one provided by a cloud provider, or [HAProxy](http://www.haproxy.org/)) with it's own name in DNS. The load balancer should be configured to forward the apiserver port (default: 6443) on the load balancer to the apiserver port on all control plane nodes. Before using a cloud provider provided load balancer, double check the doc ...
```
```suggestion
- A highly available load balancer that operates at Layer 4 (such as one provided by a cloud provider, or [HAProxy](http://www.haproxy.org/)) with its own name in DNS. The load balancer should be configured to forward the apiserver port (default: 6443) on the load balancer to the apiserver port on all control plane nodes. Before using a cloud provider provided load balancer, double check the documentation for [NAT hairpinning](https://en.wikipedia.org/wiki/Hairpinning) support.
```

zacharysarah commented:
@rcythr 👋 Any updates?

zacharysarah commented:
@rcythr 👋 Please feel free to `/reopen` when you're ready to continue.

/close
